State Management API Overhaul

drub0y opened the pull request:
Fixes #375.

drub0y commented:
NOTE: I've applied the WIP tag because this is not 100% ready to merge _yet_ as there are open discussion items that, depending on the answer, will change some of the tests that are failing right now. For example, the `FileSystemStateStore` doesn't handle some optimistic concurrency situations yet.

drub0y commented:
Ok, for now I've just bitten the bullet and implement optimistic concurrency for the `FileSystemStateStore` in a way that should work. I'd love any feedback on the approach taken. In a nutshell a specific metadata property called `@@botFramework` is added to the JSON blob that looks like this:

```JSON
"@@botFramework": {
        "rawNamespace": "test/state/namespace/9dd5d3d2-7cfe-437a-bf83-80bead6b8815",
        "rawKey": "ac6bcb6f-b990-4248-96e6-14841f7ba6e2",
        "eTag": "e1cdc90444f249eda6909ce2ac49574b"
    }
```
The original `raw[Namespace|Key]` properties are there because the file system can't support all possible characters in its paths/file names and the `eTag` property is a naked guid value that is generated anew on every write.

drub0y commented:
FYI, I'm working on rebasing against changes that went into `master` today. I need to upgrade some of the new work that uses the old state model before I can push it up otherwise the branch will be in a broken state.

tomlm commented on libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs:
```diff
@@ -50,22 +50,23 @@ public BotFrameworkAdapter(ICredentialProvider credentialProvider, RetryPolicy c
-
-            // Hand craft Claims Identity.
-            var claimsIdentity = new ClaimsIdentity(new List<Claim>
+            using (var context = new TurnContext(this, reference.GetPostToBotMessage()))
```
> using (var context = new TurnContext(this, reference.GetPostToBotMessage())) [](start = 11, length = 77)

hmmm..IDisposable is a very C# pattern.  how does this translate to node and other langs?

drub0y replied:
Yes, this is one of those .NET specific things. I would say you just don't do offer it in those languages if it doesn't make sense. 

An alternative, for something like TypeScript, would be to take another, option parameter when registering the service that is the call back you want invoked when the turn is ending.

tomlm commented on samples/AlarmBot/Startup.cs:
```diff
@@ -42,11 +43,18 @@ public void ConfigureServices(IServiceCollection services)
+
+                // Add and configure state management middleware
+                middleware.Add(new StateManagementMiddleware()
+                                .UseDefaultStorageProvider(new MemoryStateStorageProvider())
```
>UseDefaultStorageProvider [](start = 33, length = 25)

mmm.seems like SetDefaultStorageProvider() followed by .Use() 
don't feel super strongly about this one i guess

drub0y replied:
Yeah I do think the name is confusing as well. It's just that we use the `UseXXX` verb almost everywhere and `UseDefaultStorageProvider` does ultimately do what `UseStorageProvider` does. In fact it literally defers to the `UseStorageProvider` implementation under the covers, it just passes the well-known key that the system looks up the default provider by.

I think the confusing thing about the name though is that it makes it sound like you're asking to use "the" default storage provider which is not what we're saying here.

I thought about maybe `UseStorageProviderAsDefault`? Dunnuh if I love that either, but it's technically more correct.

tomlm commented on libraries/Microsoft.Bot.Builder.Core.State/StateManager.cs:
```diff
@@ -0,0 +1,143 @@
+            }
+
+            stateEntry.StateStorageEntry.SetValue(state);
+        }
```
I may not be reading this correctly, but I am concerned that 2 components that use the same state object will clobber each other.
A: var x = get(key)
A -> calls B
B: var x get(key)
B: Change x
B: Set(x)
A -> Change x (not the same X)
A: Set(x) <-- clobbers B's change
???

drub0y replied:
_Absolutely_ this is a possibility. Each state namespace, hence state manager, is a shared key space and all that contribute to it must be careful within that space not to step on one another within that space. I honestly don't think this is any different than how some other frameworks out there work honestly. For example, ASP.NET's `HttpContext::Items` collection. 

This is an important thing to call out in documentation and, for middleware especially, provide some guidance for some kind of prefixing scheme to keep their keys isolated.

Couple things though:
 1. One thing to note is that for code that is just using the `Set<T>(T value)` extension method approach, it's actually using the `Type::FullName` of the `T` and, in that case, that's the full namespace qualified name of the .NET type which would all but entirely eliminate any collision concerns.
 2. The absolutely greatest isolation is attainable by using an entirely separate state namespace. That can be done via an explicit namespace constant and via something that I just realized wasn't called out explicitly in this DCR. Developers can actually create their own strongly types state manager interfaces/classes and register those with a `UseState<TStateManager>()` instead. Those kinds of state managers are actually capable of constructing themselves from details of the current `ITurnContext`. This is actually how conversation state and user state (and eventually bot state) work under the covers. This extension point enables you create your own, completely isolated state namespace based off of dynamic values coming `ITurnContext`.

tomlm commented on libraries/Microsoft.Bot.Builder.Core.Extensions/FlexObject.cs:
```diff
@@ -1,218 +0,0 @@
-    /// <summary>
-    /// DynamicObject which seemlessly blends dictionary properties with real properties
-    /// </summary>
-    public class FlexObject : DynamicObject, ICloneable
```
>FlexObject [](start = 17, length = 10)

why do you hate my FlexObject so? :)

tomlm commented on tests/Microsoft.Bot.Builder.Prompts.Tests/ConfirmPromptTests.cs:
```diff
@@ -18,16 +19,16 @@ public class ConfirmPromptTests
-                .Use(new ConversationState<TestState>(new MemoryStorage()));
+            TestAdapter adapter = new TestAdapter();
+
+            var inPrompt = false;
```
>inPrompt [](start = 16, length = 8)

well yeah, sure... I guess you could do it this way. :smile:

tomlm commented on libraries/Microsoft.Bot.Builder.Azure/AzureTableStorageStateStorageProvider.cs:
```diff
@@ -0,0 +1,271 @@
+
+        public async Task<IStateStorageEntry> Load(string stateNamespace, string key)
+        {
+            var tableResult = await _table.ExecuteAsync(TableOperation.Retrieve(NormalizeKeyValueForTableStorage(stateNamespace), NormalizeKeyValueForTableStorage(key))).ConfigureAwait(false);
```
>NormalizeKeyValueForTableStorage(stateNamespace) [](start = 80, length = 48)

This inherently means that all Conversation state is in 1 partition which severely limits the scalability of table storage provider.  That may be OK, as maybe we should be pushing people to using blob or cosmosdb instead.  Is the namespace hierarchial? (I know that you were pushing for that in one of our meetings)  That would allow you to have blob prefix queries for blob storage or to leverage in cosmosdb for queries...

drub0y commented:
NOTE: this last push was a rebase against latest from `master`.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/ConversationStateManager.cs:
```diff
@@ -0,0 +1,30 @@
+        string ConversationId { get; }
+    }
+
+    public class ConversationStateManager : StateManager, IConversationStateManager
```
You'll have to fill me in on whatever that is. 🤔

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/ConversationStateManager.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public string ConversationId { get; }
+
+        public static string BuildStateNamespace(string channelId, string conversationId) => $"/channels/{channelId}/conversations/{conversationId}";
```
Interesting you say this. That was something I observed a while ago, but since I've been involved I've only ever seen it called "conversation state" and figured there was some history there where the channel aspect was just implied.

drub0y replied:
Null/Empty checks added.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/ConversationStateManager.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public static string BuildStateNamespace(string channelId, string conversationId) => $"/channels/{channelId}/conversations/{conversationId}";
+
+        public static string BuildStateNamespace(ITurnContext turnContext) => BuildStateNamespace(turnContext.Activity.ChannelId, turnContext.Activity.Conversation.Id);
```
Done.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/DeferredValueStateStorageEntry.cs:
```diff
@@ -0,0 +1,39 @@
+﻿namespace Microsoft.Bot.Builder.Core.State
+{
+    public abstract class DeferredValueStateStorageEntry : StateStorageEntry
```
The storage provider itself doesn't know the T as it's reading from the underlying representation. It isn't until someone attempts to read the value that the T is known. The storage providers also gain nothing by having more strongly typed instances because, at the end of the day, they all end up doing runtime type inspection to do whatever serialization they need to do anyway.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs:
```diff
@@ -0,0 +1,330 @@
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+
+namespace Microsoft.Bot.Builder.Core.State
```
Yes, that's a really good idea. I actually have some notes about how I want to change some of this implementation anyway. The basics just needs to work and an issue needs to be filed to do a deep review (and write some serious tests that target this provider specifically).

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs:
```diff
@@ -0,0 +1,330 @@
+
+        public FileSystemStateStorageProvider(string basePath)
+        {
+            _basePath = basePath;
```
Added.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs:
```diff
@@ -0,0 +1,330 @@
+
+        public Task Delete(string stateNamespace)
+        {
+            Directory.Delete(ConvertStateNamespaceToDirectoryPath(_basePath, stateNamespace));
```
There is no async version of `Delete`..... still. 😕

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs:
```diff
@@ -0,0 +1,330 @@
+        {
+            foreach (var invalidPathChar in Path.GetInvalidFileNameChars())
+            {
+                key = key.Replace(invalidPathChar, '@');
```
That's a totally valid point about collisions I would say that disallowing them in the provider itself would lead to provider specific nuances that would violate the abstraction. 

Maybe the answer is we just black list a bunch of these characters at the state manager key level? I'm totally open to better suggestions. I would suggest that it shouldn't hold up the PR. We've already agree (comment above) we _NEED_ to put in an issue in to come back and harden this entire implementation later, this could just be one of the bullet points on that issue. You're call, LMK.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs:
```diff
@@ -0,0 +1,330 @@
+            }
+        }
+
+        private sealed class FileSystemStateStorageEntry : DeferredValueStateStorageEntry
```
So, yeah, it should made internal. Done.

I believe it should be left `sealed` though and am happy to have a debate about that as long as whoever is going to argue for unsealing this type of class supplies refreshments.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/FileSystemStateStorageProvider.cs:
```diff
@@ -0,0 +1,330 @@
+                {
+                    Directory.CreateDirectory(Path.GetDirectoryName(FilePath));
+
+                    goto tryAgain;
```
Or, ya know, a proper `RetryPolicy`? 😜

I've added a counter in for now, but this is another one of those things that needs to be a bullet point on the hardening issue we put in for this.

drub0y replied:
I do want to point out that it would take some extraordinary circumstances to end up in an infinite loop because the exception handler creates the directory and if that were to fail it would throw on that call and never hit the `goto`. So the only thing that would cause an infinite loop would be some weird scenario where creating the directory succeeds, but then that very next ms somehow the directory kept getting deleted again.

Anyway, it should absolutely have a proper retry policy passed into this class for this kind of stuff, but I've got the retry counter in there for now.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/MemoryStateStorageProvider.cs:
```diff
@@ -0,0 +1,176 @@
+
+        public Task<IEnumerable<IStateStorageEntry>> Load(string partitionKey)
+        {
+            if (partitionKey == null)
```
Originally I actually purposefully wanted to allow an empty value here, but I guess that's not really a good idea. Adding empty check.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/StateManagementMiddleware.cs:
```diff
@@ -0,0 +1,416 @@
+        private Dictionary<string, IStateStorageProvider> _configuredStateStorageProviders = new Dictionary<string, IStateStorageProvider>();
+        private Dictionary<string, StateManagerConfiguration> _configuredStateManagers = new Dictionary<string, StateManagerConfiguration>();
+
+        private bool _autoLoadEnabled;
```
> For v0, I think we should punt on the auto-load-all scenarios. We should pick "Load all" or "load on demand" and just run with it. No need for the complication of multiple paths at this point in time.

Well, I don't feel like it's in your face at all, it's just an option... in fact you'd have to go look for it if you wanted it. If you just want to use `AutoLoadAll` go ahead, that's why I added it at the very top level. 

I guess I'm wondering, if I already wrote the code, short of writing some tests (which I intend to do), is there some harm in leaving the option in there for those who might want to tune their loading options more themselves??

> My inclination is "Load All / Save all, fully automatic".
Sorry, I'm unclear on this. Are you...

 1. Just re-inforcing your first comment that there should be no partial auto-load config?
 2. Suggesting that there should in fact be no separate `AutoLoadAll`/`AutoSaveAll` methods and just have one method that controls both options in unison?
 3. Suggesting there be no control offered to the user at all and we just always do it?

If 2, I can easily add a single extension method that does both (i.e. `AutoLoadAndSaveAll()`) if you feel like that would be easier than having to call two separate ones.

If 3, I would very much disagree with this and would definitely argue for continuing to offer _some_ degree of control. I think the whole auto-loading thing is a premature optimization and only through actual monitoring and tuning of specific bot implementations would someone truly be able to tell if it's worth autoloading certain state.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/StateManagementMiddleware.cs:
```diff
@@ -0,0 +1,416 @@
+        internal static string BuildStateManagerNamespaceForType(Type stateManagerType) => $"TYPED:{stateManagerType.FullName}";
+    }
+
+    public class StateManagerConfiguration
```
Yes, these were later additions that I did inline and never moved out. Will move out.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/StateManagementMiddleware.cs:
```diff
@@ -0,0 +1,416 @@
+        }
+    }
+
+    public static class StateManagementMiddlewareExtensions
```
Yep, done.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/UserStateManager.cs:
```diff
@@ -0,0 +1,23 @@
+
+    public class UserStateManager : StateManager, IUserStateManager
+    {
+        public static string PreferredStateStoreName = $"{nameof(UserStateManager)}.PreferredStateStore";
```
Leftovers from earlier implementation. Deleted.

drub0y commented on libraries/Microsoft.Bot.Builder.Core/BotAdapterServiceFactoryCollection.cs:
```diff
@@ -0,0 +1,78 @@
+
+namespace Microsoft.Bot.Builder
+{
+    public interface IBotAdapterServiceFactoryCollection : IEnumerable<KeyValuePair<Type, Func<object>>>
```
Sorry, leftovers from when I was adding top level DI into the mix very early on. Thought I deleted.... deleted now.

drub0y commented:
@cleemullins I have addressed the issues we discussed today. It looks like I need to rebase in some changes from master regarding the middleware that were made and now conflict. I will take care of that ASAP. Please take another pass over and let me know if I have sufficiently addressed all of your feedback.

Stevenic commented on libraries/Microsoft.Bot.Builder.Core.State/IStateStorageProvider.cs:
```diff
@@ -0,0 +1,40 @@
+        /// </summary>
+        /// <param name="stateNamespace"></param>
+        /// <returns></returns>
+        Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace);
```
>Task<IEnumerable<IStateStorageEntry>> Load(string stateNamespace); [](start = 8, length = 66)

I can't change the return type for an overload on the JS side so this should be renamed to something like LoadAll()

drub0y replied:
So this is about offering the _ability_ for the provider to optimize around the bulk loading scenario in some way, however the provider _might not_ be able to offer any sort of real optimizations around it. 

For example, the file system implementation essentially has to enumerate the files in the namespace folder on the file system and load each one. However, the Azure Table Storage implementation can obviously optimize this by doing one partition key query. Fundamentally I think you have to offer it as a separate concept at the

drub0y replied:
> I can't change the return type for an overload on the JS side so this should be renamed to something like LoadAll()

I can certainly do that if we feel strongly enough. I feel like this could be one of those cases where we employ the patterns that are available to/established to each ecosystem. In .NET the overload makes some sense, but... again, I'm not married to it and could just change it to a `LoadAll` if you guys think that's better.

Stevenic replied:
Just checked and Redis has no out of the box support for Tags so how would you expect Redis to implement this?

---
In reply to: [179852818](https://github.com/Microsoft/botbuilder-dotnet/pull/376#discussion_r179852818) [](ancestors = 179852818,179845861)

drub0y replied:
Replied to internal email with more details, but to have the discussion in the open there are options such as a lexing scheme that would allow us to use `ZRANGEBYLEX` or storing a list of keys for a namespace as one redis entry and then using that list to efficiently query out all the other entries in bulk.

Stevenic commented on libraries/Microsoft.Bot.Builder.Core.State/IStateStorageProvider.cs:
```diff
@@ -0,0 +1,40 @@
+
+namespace Microsoft.Bot.Builder.Core.State
+{
+    public interface IStateStorageProvider
```
>Provider [](start = 34, length = 8)

Do we REALLY need the word "Provider" at the end of this interface and therefore every class the implements it? This is overly verbose for a JS developer.

drub0y replied:
If you're asking from the .NET side, the "provider" pattern (and naming) is pretty well established there and I would suggest keeping it (though I'm not looking to die on any hill for it), but if you don't think the name helps on the JS then I don't think you need to use it. I think this is one of those things where differing for the ecosystem could make total sense.

Stevenic replied:
Just feels overly wordy for JS and StateStorage is pretty descriptive in itself

---
In reply to: [179843934](https://github.com/Microsoft/botbuilder-dotnet/pull/376#discussion_r179843934) [](ancestors = 179843934)

drub0y replied:
I gotcha, again, I think it's totally fair to drop 'Provider' in JS. I only think we need to strive for architecture/concept alignment and naming/other styles should drift towards ecosystem patterns, but that's my opinion and if you guys, as the owners, don't agree I will of course rename back to just `IStateStorage`.

P.S. When it comes to wordy, .NET ain't go nothin' Java. 😆

Stevenic commented on libraries/Microsoft.Bot.Builder.Core.State/IStateStorageProvider.cs:
```diff
@@ -0,0 +1,40 @@
+{
+    public interface IStateStorageProvider
+    {
+        IStateStorageEntry CreateNewEntry(string stateNamespace, string key);
```
>key [](start = 72, length = 3)

It's not clear if the key needs to be unique at time of creation. I'm assuming it doesn't but the documentation to clarify this.

drub0y replied:
You're right, this should be doc'd. For the record, it doesn't and because we're last write wins now it could theoretically end up overwriting any existing key that's there in the store when saved if there's any kind of concurrency going on.

Stevenic replied:
I commented it that way.

---
In reply to: [179846161](https://github.com/Microsoft/botbuilder-dotnet/pull/376#discussion_r179846161) [](ancestors = 179846161)

Stevenic commented on libraries/Microsoft.Bot.Builder.Core.State/IStateStorageProvider.cs:
```diff
@@ -0,0 +1,40 @@
+        /// <param name="stateNamespace"></param>
+        /// <param name="key"></param>
+        /// <returns></returns>
+        Task<IStateStorageEntry> Load(string stateNamespace, string key);
```
What's the expected behavior if a object is missing?

drub0y replied:
You get a `null` back in C# indicating that no storage entry exists for the specified namespace/key. So I guess for JS you can decide if `null` or `undefined` makes more sense.

Stevenic commented on libraries/Microsoft.Bot.Builder.Core.State/IStateStorageEntry.cs:
```diff
@@ -0,0 +1,13 @@
+
+        T GetValue<T>() where T : class, new();
+
+        void SetValue<T>(T value) where T : class;
```
>void SetValue<T>(T value) where T : class; [](start = 8, length = 42)

I'm only implementing the getValue() for JS. And I shortened it to get() as that's what a JS developer expects.

drub0y replied:
Yep, catering to the JS patterns here makes sense to me.

drub0y commented on libraries/Microsoft.Bot.Builder.Core.State/IStateStorageProvider.cs:
```diff
@@ -0,0 +1,40 @@
+
+        Task Save(IEnumerable<IStateStorageEntry> stateStorageEntries);
+
+        Task Delete(string stateNamespace);
```
As stated above, it's about offering the ability for optimization for bulk operation _if_ the provider can handle it, otherwise the provider emulates as it needs to... in this case maybe just calling the singular version for each one in a loop.

drub0y commented:
Rebased against latest from `master` again.

drub0y commented:
Just rebased against latest from `master` again.

cleemullins commented:
After quite a bit of internal discussion, the more recent State changes are our current plan of record. This PR may come back in the future, but for now I'm going to close it out.
